internal/ui/backup/progress.go

```diff
@@ -84,191 +82,165 @@ func NewProgress(printer ProgressPrinter, interval time.Duration) *Progress {
-			forceUpdate = true
+			now = time.Now()
 		}
 
-		// limit update frequency
-		if !forceUpdate && (p.interval == 0 || time.Since(lastUpdate) < p.interval) {
+		p.mu.Lock()
+		state := p.state
+
+		if state&stateStarted == 0 {
+			p.mu.Unlock()
 			continue
 		}
-		lastUpdate = time.Now()
 
-		p.printer.Update(total, processed, errors, currentFiles, p.start, secondsRemaining)
+		var secondsRemaining uint64
+		if state&stateFinished != 0 {
+			secs := float64(now.Sub(p.start) / time.Second)
+			todo := float64(p.total.Bytes - p.processed.Bytes)
+			secondsRemaining = uint64(secs / float64(p.processed.Bytes) * todo)
+		}
+
+		p.printer.Update(p.total, p.processed, p.errors, p.currentFiles, p.start, secondsRemaining)
+		p.mu.Unlock()
 	}
 }
 
 // Error is the error callback function for the archiver, it prints the error and returns nil.
 func (p *Progress) Error(item string, err error) error {
-	cbErr := p.printer.Error(item, err)
+	p.mu.Lock()
+	p.errors++
+	p.state |= stateStarted
+	p.mu.Unlock()
 
-	select {
-	case p.errCh <- struct{}{}:
-	case <-p.closed:
-	}
-	return cbErr
+	return p.printer.Error(item, err)
 }
 
 // StartFile is called when a file is being processed by a worker.
 func (p *Progress) StartFile(filename string) {
-	select {
-	case p.workerCh <- fileWorkerMessage{filename: filename}:
-	case <-p.closed:
-	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.currentFiles[filename] = struct{}{}
+}
+
+func (p *Progress) addProcessed(c Counter) {
+	p.processed.Files += c.Files
+	p.processed.Dirs += c.Dirs
+	p.processed.Bytes += c.Bytes
+	p.state |= stateStarted
 }
 
 // CompleteBlob is called for all saved blobs for files.
 func (p *Progress) CompleteBlob(bytes uint64) {
-	select {
-	case p.processedCh <- Counter{Bytes: bytes}:
-	case <-p.closed:
-	}
+	p.mu.Lock()
+	p.addProcessed(Counter{Bytes: bytes})
+	p.mu.Unlock()
 }
 
 // CompleteItem is the status callback function for the archiver when a
 // file/dir has been saved successfully.
 func (p *Progress) CompleteItem(item string, previous, current *restic.Node, s archiver.ItemStats, d time.Duration) {
-	p.summary.Lock()
+	p.mu.Lock()
 	p.summary.ItemStats.Add(s)
 
 	// for the last item "/", current is nil
 	if current != nil {
 		p.summary.ProcessedBytes += current.Size
 	}
 
-	p.summary.Unlock()
+	p.mu.Unlock()
 
 	if current == nil {
 		// error occurred, tell the status display to remove the line
-		select {
-		case p.workerCh <- fileWorkerMessage{filename: item, done: true}:
-		case <-p.closed:
-		}
+		p.mu.Lock()
+		delete(p.currentFiles, item)
+		p.mu.Unlock()
 		return
 	}
 
 	switch current.Type {
 	case "dir":
-		select {
-		case p.processedCh <- Counter{Dirs: 1}:
-		case <-p.closed:
-		}
+		p.mu.Lock()
+		p.addProcessed(Counter{Dirs: 1})
+		p.mu.Unlock()
 
 		if previous == nil {
 			p.printer.CompleteItem("dir new", item, previous, current, s, d)
-			p.summary.Lock()
+			p.mu.Lock()
 			p.summary.Dirs.New++
-			p.summary.Unlock()
+			p.mu.Unlock()
 			return
 		}
```

Comment:
what about removing this `return` by replacing the next `if` with `else if`? That would make the control flow easier to follow.

Reply:
Changed to a three-case switch statement.